advection in AD-Mode

mjlosch opened the pull request:
## What changes does this PR introduce?
Feature update, maybe a small bug fix (in seaice_lsr)

## What is the current behaviour? 
with `inAdExact = .False.,` one can trigger a number of approximations of the adjoint, some are related to KPP, some to SEAICE, some are related to advection, but only with advection scheme 33 and without multiDimAdvection. I find this confusing and misleading, but I would like to use the advection functionality separately. Also the flags `inAdTrue` and `inAdFalse` are not intuitive, as `inAdTrue = .FALSE.` is possible.

## What is the new behaviour 
A new parameter `useApproxAdvectionInAdMode` allows to just use the advection trick without unclear implications in other parts of the code: use advScheme 33 in forward part of the simulation but 30 in the AD-part, now also implemented for seaice_advection and gad_advection (i.e. with multiDimAdvection enabled). The flags `inAdTrue` and `inAdFalse` have been removed. Instead, `inAdMode` is hardwired to be `T` and `F` by S/R autodiff_inadmode_set/unset_ad. This requires that the new flag `useApproxAdvectionInAdMode` and also `inAdExact` need to be used together with `inAdMode` to have the correct effect.

## Does this PR introduce a breaking change? 
Old set ups are not affected, except for the combination of `inAdExact = .False.,` and advection scheme 33 with multidim-advection (which did not have any effect before this change, but now it does!). I removed the `inAdMode` if-statement and TAF-related code from seaice_lsr, but that shouldn't have any effect on any setup either, except that now the maximum number of iterations is the same in forward and AD mode. This made it necessary to adjust some parameters in verification experiments.

## Suggested addition to `tag-index`
o pkg/autodiff: add new parameter useApproxAdvectionInAdMode to replace inAdMode
- if useApproxAdvectionInAdMode is True and advection scheme 33 use
  advection scheme 30 in AD-model in seaice_advection, gad_advection
  and gad_calc_rsh, model stops when parameters are set inconsistently
- defaults to False, but is reset to True if inAdExact = False to be
  backwards compatible
- important change in the meaning of variable inAdMode: now it is always true after 
  s/r autodiff_inadmode_set_ad and always false after s/r autodiff_inadmode_unset_ad. 
  The non-intuitive flags inAdTrue and inAdFalse, which could be T or F, have been 
  removed from the code.
- remove unnecessary TAF related and inAdMode code from seaice_lsr.F, 
  adjust SEAICElinearIterMax = SOLV_MAX_FIXED to reproduce old results after this change
- change global_ocean.seaice_ad to test this new flag and functionality, gradient checks 
  necessarily are poorer after that because advection AD is approximated

mjlosch commented:
@jm-c haven't done this yet, but should we test this functionality in one of the verification experiments? I suggest global_ocean.cs32x15.seaice

mjlosch commented:
After a month of contemplation, I realised that one needs to change the way `inAdMode` is used for this trick. Since I want to differentiate between `inAdExact = .FALSE.`, and `useApproxAdvectionInAdModel = .TRUE.`, I always have to use `inAdMode` together with either of the flags. This means, that configurations that are using `inAdMode` in unpublished code will be affected by this change.
It also means that `inAdMode` could be hardwired to be True and False in autodiff_inadmode_set/unset_ad.F and we could get rid of the confusing flags inAdTrue/False (which, in theory, can be either True or False), or at least we could hardwire these flags in autodiff_readparms.F independently of the value of `inAdExact` or `useApproxAdvectionInAdModel`. Any opinions on that?

mjlosch commented:
@jm-c 
could you please update the reference output for global_ocean.cs32x15/input_ad.seaice?

timothyas commented:
Just mentioning that I'm not ignoring the review request, and I should be able to get to it between now and next Wednesday.

jm-c commented:
@mjlosch The changes you suggest for inAdMode (and removing inAdTrue, inAdFalse) might make the line of code longer
but it's certainly much easier to follow and understand. I would vote for.

jm-c commented on pkg/autodiff/autodiff_readparms.F:
```diff
@@ -98,13 +107,31 @@ SUBROUTINE AUTODIFF_READPARMS( myThid )
+       WRITE(msgBuf,'(A)') '**WARNING** AUTODIFF_READPARMS: '//
+     &      'resetting useApproxAdvectionInAdMode=.FALSE., '//
+     &      'because inAdExact=.FALSE.'
+       CALL PRINT_MESSAGE( msgBuf, errorMessageUnit,
```
I found this warning message not very clear:  I personally prefer to see what is the new value a specific parameter
is reset to rather than the previous value it use to have before the reset.

mjlosch replied:
OK, fixed

jm-c commented on pkg/generic_advdiff/gad_advection.F:
```diff
@@ -615,16 +626,26 @@ SUBROUTINE GAD_ADVECTION(
      O             af, myThid )
 #ifndef ALLOW_AUTODIFF
-        ELSEIF (advectionScheme.EQ.ENUM_OS7MP ) THEN
+        ELSEIF (advectionScheme.EQ.ENUM_OS7MP) THEN
```
Minor comment here regarding space in these Advection-scheme test value: If the spacing is changed, might be a good
idea to change it the same way in sibling routine gad_calc_rhs.F (were several lines have changed also for the same reason). I personally like the spacing you put in gad_calc_rhs.F, but it's a minor point, a consistent change would be better.

mjlosch replied:
OK

jm-c reviewed (commented):
I just wanted to add some comments (and not  to write a "review"), but just clicked on the wrong button ...

jm-c commented:
@mjlosch I updated the 2 (adm & tlm) reference output for global_ocean.cs32x15.seaice. AD gradient check are not as good as
they were before, but I guess this is expected ? And for TLM, it's a bit better.

mjlosch commented:
thanks @jm-c for the comments. I implemented the requested changes and resolved the conversations. The AD gradient checks should become worse, because we are no approximated the AD code on purpose. Not sure why the TLM checks get better, maybe because advScheme 33 is a bit more stable than 30?

I also implemented my suggestion of removing inAdFalse/True. I think it's an improvement, but it should also be documented in the tag-index (minimum), because it changes the meaning of inAdMode. Now it always means what the name implies (i.e. if True, we are "in AD-mode" in the backwards sweep, if False, we are in forward mode), and I tried to document that in AUTODIFF_PARAMS.h. Before this change inAdMode could be false, even when we were "in AD-mode", and it meant that we don't need to apply the modifications for a more stable (and inexact) AD run.

jm-c commented:
@mjlosch I have an other comment regarding the few check & stop in autodiff_readparms.F related to useApproxAdvectionInAdMode
and advection schemes: For my point of view (and I think the code allows this) I might want to use this feature (useApproxAdvectionInAdMode) for all tracers that use advScheme=33, even if other advection schemes are
used for other tracers/ice fields. A typical (and frequent) example would be SEAICEadvScheme=33 with tempAdvScheme=saltAdvScheme=2 (or = an other linear advection scheme). But the current check would stop the model run for
this type of param setting.

mjlosch commented:
@jm-c 
good point. What do you suggest?

1. We can silently ignore it if a parameter combination makes `useApproxAdvectionInAdMode` useless
2. We can just reduce this stop to a warning.
3. We can require that at least one of the advection schemes (and there are 3+ parameters for advection in pkg/seaice alone) is 33 if `useApproxAdvectionInAdMode=.TRUE.`
4. I am sure you have more suggestions (-;


I would favour the warning (2).

jm-c commented:
I think (3) would be better, i.e., a warning or a stop if ENUM_DST3_FLUX_LIMIT (=33) is never used
and useApproxAdvectionInAdMode=T (because, in this case, setting useApproxAdvectionInAdMode=T is 
useless). And if we anticipate that a rarely use advected seaice-field might not be accounted for
in this check, then a warning would be better.
On Mon, Jan 25, 2021 at 07:33:56AM -0800, Martin Losch wrote:
> @jm-c 
> good point. What do you suggest?
> 
> 1. We can silently ignore it if a parameter combination makes `useApproxAdvectionInAdMode` useless
> 2. We can just reduce this stop to a warning.
> 3. We can require that at least one of the advection schemes (and there are 3+ parameters for advection in pkg/seaice alone) is 33 if `useApproxAdvectionInAdMode=.TRUE.`
> 4. I am sure you have more suggestions (-;
> 
> 
> I would favour the warning (2).
> 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/MITgcm/MITgcm/pull/404#issuecomment-766899746

mjlosch commented:
OK, I went with option (3). I agree, that it does make the most sense.

timothyas commented on pkg/seaice/seaice_lsr.F:
```diff
@@ -664,11 +663,7 @@ SUBROUTINE SEAICE_LSR( myTime, myIter, myThid )
-      ELSE
-        SOLV_MAX_TMP = linearIterLoc
-      ENDIF
+      SOLV_MAX_TMP = SOLV_MAX_FIXED
```
This is the only possible inconsistency I can find, but I'm not familiar with this part of the code. It seems that old behavior used to use a different value for `SOLV_MAX_TMP` when `inAdMode=.TRUE.`, is this taken care of somewhere else I missed?

mjlosch replied:
@timothyas
I silently removed this part, because the LSR-adjoint does not work anyway and I didn't see the point of this code: why should the maximum number if iterations be different in Ad-mode than in FWD-mode? What's important here is that SOLV_MAX_TMP does not exceed SOLV_MAX_FIXED, because that determines the length of the tapes (from the_main_loop.F):
```
CADJ INIT comlev1_lsr = COMMON,nchklev_1*MPSEUDOTIMESTEPS*SOLV_MAX_FIXED
CADJ INIT comlev1_bibj_lsr = COMMON,
CADJ &  nchklev_1*nSx*nSy*nthreads_chkpt*MPSEUDOTIMESTEPS*SOLV_MAX_FIXED
```
We can totally get rid of this code (i.e. get rid of `SOLV_MAX_TMP`, replace it by `linearIterLoc` everywhere, and check that `SOLV_MAX_FIXED` is never exceeded, similar to `numItersMax` in cg2d_nsa.F, maybe that's the better solution. What do you think?

timothyas replied:
That makes the most sense to me, getting rid of the code and just replacing it by `linearIterLoc`. I agree, I don't understand why the adjoint would need a different maximum number of iterations, and this is more straightforward.

mjlosch replied:
@timothyas 
please have a look at my latest commit where I implemented this idea.

timothyas replied:
Looks good to me!

timothyas commented on pkg/kpp/kpp_routines.F:
```diff
@@ -1085,7 +1085,7 @@ subroutine Ri_iwmix (
 #else /* EXCLUDE_KPP_SHEAR_MIX */
 # ifdef ALLOW_AUTODIFF
-            if ( inAdMode ) then
+            if ( inAdMode .AND. .NOT. inAdExact ) then
```
It's a bit surprising this wasn't already here! The adjoint keeps us on our toes I guess... Good fix.

timothyas reviewed (commented):
Just have one question regarding `seaice_lsr` (which I made inline) but otherwise this looks good to me

timothyas reviewed (approved):
@jm-c may have some final requests, but this looks good to me

mjlosch commented:
@jm-c can you confirm, that all reference AD/TLM-experiments pass on your reference machine? On my linux box I temporarily found that there are only 3 digits of agreement for global_ocean.cs32x15.seaice even directly after 1f8336b, but I cannot reproduce that now (i.e. now everything looks fine to me). git tricked me again ...

jm-c commented:
Checked on villon, no problem, got 16 digits for global_ocean.cs32x15.seaice in both ADM & TLM

jm-c commented:
One thing that might be useful: In gad_advection.F (and other similar S/R), if we rename the 2 arguments: advectionScheme &vertAdvecScheme to something else (e.g. advSchemeArg) and copy them to advectionScheme & vertAdvecScheme , we can
apply the useApproxAdvectionInAdMode trick to the copy and then simplify a bit the rest of the routine.
Also this way, if we decide an other non-linear advection scheme (e.g., 7 or 77) will be replaced in AdMode by an other one (e.g., 20 or 30 again), it will be straightforward.

mjlosch commented:
@jm-c that sounds like a very useful suggestion, but I am afraid that I don't quite understand how you want to do it. Like this:
```
        advectionScheme = advectionSchArg
#ifdef ALLOW_AUTODIFF
        IF ( useApproxAdvectionInAdMode ) THEN
         IF (advectionSchArg.EQ.ENUM_DST3_FLUX_LIMIT )
     &          advectionScheme = ENUM_DST3
C        here is room for more of these tricks:
         ...
       ENDIF
#endif /* ALLOW_AUTODIFF */
```

(similar for vertAdvecScheme) and then remove the autodiff part later on?

jm-c commented:
@mjlosch Yes, you got it, it's exactly as I suggested. Also, one little advantage of doing this way is that it's more clear that we
really switch to using an alternate advection scheme in Adjoint mode (and not just part of it, e.g., in 1 direction but not in other directions).

mjlosch commented:
@jm-c I really like your suggestion! Please have a look. I think that this is now good to go.

jm-c commented on pkg/generic_advdiff/gad_advection.F:
```diff
@@ -190,6 +196,18 @@ SUBROUTINE GAD_ADVECTION(
+C     combination is possible.
+       IF ( advectionSchArg.EQ.ENUM_DST3_FLUX_LIMIT )
+     &      advectionScheme = ENUM_DST3
+       IF ( vertAdvecScheme.EQ.ENUM_DST3_FLUX_LIMIT )
```
A very minor suggestion: why not testing (line 207) for vertAdvecSchArg.EQ. ... instead of vertAdvecScheme,
the same way it tests for advectionSchArg (and not advectionScheme) ? Processing the 2 advection-schemes in the
same and symetric way seems better to me. I think the same comment also applies to gad_calc_rhs.F

mjlosch replied:
thanks, @jm-c I'll correct that.

jm-c commented:
@mjlosch I am going to make a small PR on verification_other to update few files there (e.g., some data.seaice);
this is the reason why I've just merged-in master here, so that it would be easier to have both repos in sync.

jm-c reviewed (approved):
LGTM I will check one more time the new check+stop/warning but code changes look good
and with my latest tweak I don't think I need any update in verification_other experiments.
One last thing: suggest to add one more comment to tag-index related to upgrade of AD experiment global_ocean.cs32x15.seaice.

mjlosch commented on verification/offline_exf_seaice/input_ad/data.seaice:
```diff
@@ -26,7 +26,7 @@
  SEAICEuseDYNAMICS  =.FALSE.,
  LSR_ERROR          = 1.E-12,
- SEAICElinearIterMax= 1500,
+ SEAICElinearIterMax= 500,
```
@jm-c you didn't change this back on purpose (after you latest commit `SEAICElinearIterMax` defaults to `SOLV_MAX_FIXED = 500`)? But maybe good to have it here explicitly, so that we don't forget?

jm-c replied:
Yes, I kept this one since originally SEAICElinearIterMax was specified there, so prefer to keep the updated file (with the updated value). And for all the other data.seaice that did not have SEAICElinearIterMax set-up initially, I feel there was no
need to keep the update (that just set-it to the new default).

jm-c commented:
@mjlosch A minor comment regarding useApproxAdvectionInAdMode check in autodiff_read_params.F : the logic is good. the choice between a stop or a warning is a bit subjective: when pkg/generic_advdiff is not compiled, a stop might look better than a warning ;
and the stop in case ENUM_DST3_FLUX_LIMIT is not used coould have been a warning (but we can keep a stop). Anyway, if you feel making a small changes there let me know. Otherwise I will move forward to merge this PR.

mjlosch commented:
@jm-c I have not strong feelings about the stop/warning. We can do it as you suggest. Please feel free to do it, as I don't want to screw things up in the last minute.

jm-c commented:
@mjlosch I only changed 1 warning to a stop, but left the other as full stop. Going to merge it later today.
